Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +182,83 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts, eval, evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Maps
+
+variable {α' σ' : Type*} (f : α' → α) (g : σ ≃ σ')
+
+/--
+`M.comap f` pulls back the alphabet of `M` along `f`. In other words, it applies `f` to the input
+before passing it to `M`.
+-/
+@[simps]
+def comap : DFA α' σ where
+  step s a := M.step s (f a)
+  start := M.start
+  accept := M.accept
+
+@[simp]
+theorem comap_id : M.comap id = M := rfl
+
+@[simp]
+theorem evalFrom_comap (s : σ) (x : List α') :
+    (M.comap f).evalFrom s x = M.evalFrom s (x.map f) := by
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+@[simp]
+theorem eval_comap (x : List α') : (M.comap f).eval x = M.eval (x.map f) := by simp [eval]
+
+@[simp]
+theorem accepts_comap : (M.comap f).accepts = List.map f ⁻¹' M.accepts := by
+  ext x
+  conv =>
+    rhs
+    rw [Set.mem_preimage, mem_accepts]
+  simp [mem_accepts]
+
+/-- Lifts an equivalence on states to an equivalence on DFAs. -/
+@[simps apply_step apply_start apply_accept]
+def reindex : DFA α σ ≃ DFA α σ' where
```

Comment:
I am a bit suspicious that somebody might later need a more general version than `σ ≃ σ'` but I don't mind keeping this definition as it is.

Reply:
What would the more general version look like?

Reply:
In my project, I ended up needing this construction:
https://github.com/madvorak/chomsky/blob/a27e2145c36b0a96faf405cb415d020dae97b04a/Grammars/Classes/Unrestricted/Basics/Lifting.lean#L35

The explanation (but written for Lean 3) is here:
https://arxiv.org/abs/2302.06420 (page 6)

I don't have a particular argument why something similar should be useful for DFAs, just a gut feeling.

Reply:
The analogy is a bit obfuscated by the fact that what you call `σ` is bundled into the structure in my project.

Reply:
I guess it might come handy when proving that regular expressions and DFAs are equivalent (TODO in `RegularExpressions.lean`).

Reply:
However, it would probably be done on the `NFA` structure, so irrelevant for now I guess.